### ISSUE TITLE
Deprecate homomorphic mapped type over primitive

### ISSUE THREAD

**Issue by rotu** [microsoft/TypeScript#56940]
### 🔍 Search Terms

homomorphic mapped primitive

### ✅ Viability Checklist

- [X] This wouldn't be a breaking change in existing TypeScript/JavaScript code
- [X] This wouldn't change the runtime behavior of existing JavaScript code
- [X] This could be implemented without emitting different JS based on the types of the expressions
- [X] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
- [X] This isn't a request to add a new utility type: https://github.com/microsoft/TypeScript/wiki/No-New-Utility-Types
- [X] This feature would agree with the rest of our Design Goals: https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals

### ⭐ Suggestion

I would like homomorphic mapped types over primitives to emit a warning if the base type is not an object type. An explicit conditional `type Foo<T> = T extends object ? { [K in keyof T]: ...} : T` can be used to recover the pre-existing behavior. This should be mentioned in the warning message.

Homomorphic mapped type over primitives has surprising behavior that bears mention in the FAQ:

> - This mapped type returns a primitive type, not an object type.
>   - Mapped types declared as `{ [ K in keyof T ]: U }` where `T` is a type parameter are known as homomorphic mapped types, which means that the mapped type is a structure preserving function of T`. When type parameter `T` is instantiated with a primitive type the mapped type evaluates to the same primitive.

This behavior is surprising since:
1. It is *not* referentially transparent; that is, replacing a type parameter with its value does not result in the same type
2. Although a mapped type *may* be assignable to the original type, it may specifically *not* be, e.g. `{[k in keyof T]: never}` whereas it always is in the case of primitives.
3. The expression `keyof T` has surprising behavior on primitive types, differing substantially from object types.
4. The syntax of a mapped type visually suggests an object-like type, which is incorrect when mapping over `null`.

### 📃 Motivating Example

```tsx
type NullValued<X, Y> = {[k in keyof X]: null}[;
type Okay = NullValued<{x:1, y:2}>; // {x:null, y:null}
type Oops = NullValued<number>; // oops! This is `number`, which is likely not what you expected
```

### 💻 Use Cases

1. What do you want to use this for?
2. What shortcomings exist with current approaches?
  - You have to know about homeomorphic mapped types and anticipate this corner case.
3. What workarounds are you using in the meantime?
  - A type guard `type Foo<T extends object> = { [K in keyof T]: ...}` guards against the current behavior.
  - A conditional `type Foo<T> = T extends object ? { [K in keyof T]: ...} : T` makes the current behavior explicit.

**Comment by rotu** [comment]
FYI @Andarist this seems up your alley and I think you'd know if it's possible.

**Comment by RyanCavanaugh** [comment]
It's currently the rule that instantiation is not allowed to produce errors. This is super important because instantiations that don't end up in any user-space type can and do occur, and it's not straightforward to track which side effects (errors) came from which instantiations.

It seems like just telling people "If you have a HMT, put an `extends object` constraint on it" is some concrete guidance that would actually be pretty easy to write into a lint rule (since HMT-ness is syntactic)

**Comment by fatcerberus** [comment]
> It is *not* referentially transparent; that is, replacing a type parameter with its value does not result in the same type

To be fair, this is true of distributive conditional types too.  In both cases the rule is consistent: the type only distributes and/or is homomorphic over a naked type parameter.

**Comment by Andarist** [comment]
Note that the distributivity of conditional types is not exactly syntactic:
```ts
type Test<T> = [T][0] extends any ? [T] : never;
type Result = Test<"a" | "b">; // ["a"] | ["b"]
```

The type parameter has to be naked but it can become naked after simplifications.

**Comment by rotu** [comment]
> It's currently the rule that instantiation is not allowed to produce errors. This is super important because instantiations that don't end up in any user-space type can and do occur, and it's not straightforward to track which side effects (errors) came from which instantiations.

Hmm... that is a bit hairy and I don't have a good answer.

> It seems like just telling people "If you have a HMT, put an `extends object` constraint on it" is some concrete guidance that would actually be pretty easy to write into a lint rule (since HMT-ness is syntactic)

The `extends` check is not ideal. There are still a number of corner cases that sneak by:

`never`, `object`, `{}`, `any` are all such examples.

Additionally, `object`, `Object`, `{}`, `any` all extend each other so an `extends` guard doesn't really make sure that if it looks like HMT it *is* HMT.

---

> To be fair, this is true of distributive conditional types too. In both cases the rule is consistent: the type only distributes and/or is homomorphic over a naked type parameter.

It's confusing there too, but it seems like it could at least be useful sometimes. I would love to see that be more consistent eventually, too!

I'm still trying to wrap my head around how these are considered distributivity of any sort:

```ts
type IsOne<X> = X extends 1 ? true : false
type Z1 = IsOne<1 | never> // true
type Z2 = IsOne<1 | 2> // boolean
type Z3 = IsOne<1 | number> // false
```

**Comment by fatcerberus** [comment]
Re: The distributive examples: `IsOne<T | U>` should be read as `IsOne<T> | IsOne<U>` - it’s like an array.map().

* `Z1` - `never` gets eagerly collapsed out of unions, so that’s just `IsOne<1>` and the result is self-evident.  The result would be the same if not, however - `never` extends everything so you’d get `true | true` which collapses to `true`
* `Z2` - `1 extends 1` is true but `2 extends 1` is false so you get `true | false` AKA `Boolean`
* `Z3` - `1 | number` eagerly collapses to `number` so you just get `IsOne<number>` which is false.

All told, it seems like maybe you’re struggling more with the union reduction rules than the distributive behavior itself (which at the end of the day is just “map over unions”)

**Comment by rotu** [comment]
Yes, that is how it works out case-wise. If this were truly "distributive" in the set-theoretic sense, then `f(a | b)` should be exactly `f(a) | f(b)`. That is, `IsOne<number>` would have to be `boolean` to restore distributivity!

**Comment by fatcerberus** [comment]
Indeed, but TS can't represent infinite unions, so in practice you can't distribute over `number` (or `string`, or `object`, etc.).

What is it about TS's type system that invites such interest in the mathematics of it all?  I ultimately got into category theory because of TS and I'm really beginning to think I'm not the only one...

**Comment by rotu** [comment]
> > It's currently the rule that instantiation is not allowed to produce errors. This is super important because instantiations that don't end up in any user-space type can and do occur, and it's not straightforward to track which side effects (errors) came from which instantiations.

> Hmm... that is a bit hairy and I don't have a good answer.

Actually I came up with a precedent: recurrence.

```ts
type Depth<X> = X extends any ? Depth<X> : never
type z = Depth<1> // error TS2589: Type instantiation is excessively deep and possibly infinite.
```

---

> Indeed, but TS can't represent infinite unions, so in practice you can't distribute over `number` (or `string`, or `object`, etc.).

I think you've captured the design rationale in that statement. The type `string` *represents* the union of all strings but *is not* a "union type". To give another example, `{x:1}|{x:2}` represent the same possible values as `{x:1|2}` but because being a "union type" is visible, optimizing out the difference would be a breaking change!

> What is it about TS's type system that invites such interest in the mathematics of it all? I ultimately got into category theory because of TS and I'm really beginning to think I'm not the only one...

I think that it's very simple. JavaScript is a complex monster of a programming language. TypeScript uses general, abstract reasoning to tame that beast. So if you tend to have a mind for abstractions, you wind up interested in both.

**Comment by fatcerberus** [comment]
> I think you've captured the design rationale in that statement. The type string represents the union of all strings but is not a "union type". To give another example, {x:1}|{x:2} represent the same possible values as {x:1|2} but because being a "union type" is visible, optimizing out the difference would be a breaking change!

Indeed, you've got it.  In set theory, set union is a monoid operation and just gives you another set, mechanically the same as any other.  Same for set intersection.  In TS, there are separate concepts for union and intersection types.  That can trip you up if you approach things purely set-theoretically, particularly in cases where the distinction matters (like distributivity).
